kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/ImmutableCollection.kt

```diff
@@ -16,24 +16,76 @@
 
 package kotlinx.collections.immutable
 
+/**
+ * A generic immutable collection of elements. Methods in this interface support only read-only access to the collection;
+ * modification operations are supported through the [PersistentCollection] interface.
+ *
+ * Implementors of this interface take responsibility to be immutable,
+ * that is objects reachable from an instance of the implementor should never be mutated.
```

Comment:
We don't require all reachable objects, e.g., collection elements, to be fully immutable.

Reply:
Rewrote to "that is objects defining structure of an instance of the implementor should never be mutated."

Alternatives are welcomed, I found it difficult to describe immutability in few words.

Reply:
Ok.

I was trying to write that immutable collections implementors must be thread-safe.
Statement "_Once constructed they must contain/retain the same elements in the same order_" doesn't demand thread-safety, as it doesn't prevent mutating underlying structure in a way that elements do not change order and no element get removed or added.

But should we demand thread-safety?